File: android/src/main/kotlin/com/example/audio_waveform_recorder/AudioWaveformPlugin.kt
package com.example.audio_waveform_recorder

import android.Manifest
import android.content.pm.PackageManager
import android.media.*
import android.os.Build
import androidx.core.content.ContextCompat
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.abs
import kotlin.math.sqrt

class AudioWaveformPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private var context: android.content.Context? = null

    // Recording
    private var mediaRecorder: MediaRecorder? = null
    private var currentFilePath: String? = null
    private var recordingStartTime: Long = 0L

    // Playback
    private var mediaPlayer: MediaPlayer? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "audio_waveform_recorder")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        mediaRecorder?.release()
        mediaPlayer?.release()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {

            // ── Permissions ──────────────────────────────────────────────
            "hasMicPermission" -> {
                val granted = ContextCompat.checkSelfPermission(
                    context!!,
                    Manifest.permission.RECORD_AUDIO
                ) == PackageManager.PERMISSION_GRANTED
                result.success(granted)
            }

            "requestMicPermission" -> {
                // Permission request must be done from Activity context.
                // Return current state — caller should use permission_handler pkg or
                // request from their Activity for full flow.
                val granted = ContextCompat.checkSelfPermission(
                    context!!,
                    Manifest.permission.RECORD_AUDIO
                ) == PackageManager.PERMISSION_GRANTED
                result.success(granted)
            }

            // ── Recording ────────────────────────────────────────────────
            "startRecording" -> {
                val format    = call.argument<String>("format") ?: "m4a"
                val sampleRate = call.argument<Int>("sampleRate") ?: 44100
                val bitRate    = call.argument<Int>("bitRate") ?: 128000
                val channels   = call.argument<Int>("channels") ?: 1
                val outputDir  = call.argument<String>("outputDir")
                val fileName   = call.argument<String>("fileName")

                try {
                    val path = buildOutputPath(format, outputDir, fileName)
                    startMediaRecorder(path, format, sampleRate, bitRate, channels)
                    currentFilePath = path
                    recordingStartTime = System.currentTimeMillis()
                    result.success(path)
                } catch (e: Exception) {
                    result.error("RECORD_ERROR", e.message, null)
                }
            }

            "pauseRecording" -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    try { mediaRecorder?.pause(); result.success(null) }
                    catch (e: Exception) { result.error("PAUSE_ERROR", e.message, null) }
                } else {
                    result.error("UNSUPPORTED", "Pause requires Android 7+", null)
                }
            }

            "resumeRecording" -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    try { mediaRecorder?.resume(); result.success(null) }
                    catch (e: Exception) { result.error("RESUME_ERROR", e.message, null) }
                } else {
                    result.error("UNSUPPORTED", "Resume requires Android 7+", null)
                }
            }

            "stopRecording" -> {
                try {
                    mediaRecorder?.stop()
                    mediaRecorder?.release()
                    mediaRecorder = null

                    val path = currentFilePath ?: ""
                    val file = File(path)
                    val durationMs = System.currentTimeMillis() - recordingStartTime

                    result.success(mapOf(
                        "path"       to path,
                        "durationMs" to durationMs.toInt(),
                        "sizeBytes"  to file.length().toInt(),
                    ))
                    currentFilePath = null
                } catch (e: Exception) {
                    result.error("STOP_ERROR", e.message, null)
                }
            }

            "cancelRecording" -> {
                try {
                    mediaRecorder?.stop()
                } catch (_) {}
                mediaRecorder?.release()
                mediaRecorder = null
                currentFilePath?.let { File(it).delete() }
                currentFilePath = null
                result.success(null)
            }

            "getAmplitude" -> {
                val recorder = mediaRecorder
                if (recorder == null) {
                    result.success(0.0)
                    return
                }
                try {
                    // maxAmplitude returns 0-32767
                    val raw = recorder.maxAmplitude
                    val normalised = raw / 32767.0
                    result.success(normalised)
                } catch (_) {
                    result.success(0.0)
                }
            }

            // ── Playback ─────────────────────────────────────────────────
            "loadAudio" -> {
                val path = call.argument<String>("path") ?: return result.error("ARGS", "path required", null)
                try {
                    mediaPlayer?.release()
                    mediaPlayer = MediaPlayer().apply {
                        setDataSource(path)
                        prepare()
                    }
                    result.success(mapOf("durationMs" to (mediaPlayer?.duration ?: 0)))
                } catch (e: Exception) {
                    result.error("LOAD_ERROR", e.message, null)
                }
            }

            "playAudio" -> {
                try { mediaPlayer?.start(); result.success(null) }
                catch (e: Exception) { result.error("PLAY_ERROR", e.message, null) }
            }

            "pauseAudio" -> {
                try { mediaPlayer?.pause(); result.success(null) }
                catch (e: Exception) { result.error("PAUSE_ERROR", e.message, null) }
            }

            "stopAudio" -> {
                try {
                    mediaPlayer?.stop()
                    mediaPlayer?.prepare() // reset to beginning
                    result.success(null)
                } catch (_) { result.success(null) }
            }

            "seekTo" -> {
                val ms = call.argument<Int>("positionMs") ?: 0
                try { mediaPlayer?.seekTo(ms); result.success(null) }
                catch (e: Exception) { result.error("SEEK_ERROR", e.message, null) }
            }

            "getPlaybackPosition" -> {
                result.success(mediaPlayer?.currentPosition ?: 0)
            }

            "setVolume" -> {
                val vol = (call.argument<Double>("volume") ?: 1.0).toFloat()
                mediaPlayer?.setVolume(vol, vol)
                result.success(null)
            }

            "setSpeed" -> {
                val speed = (call.argument<Double>("speed") ?: 1.0).toFloat()
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    mediaPlayer?.playbackParams = mediaPlayer?.playbackParams?.setSpeed(speed)
                        ?: PlaybackParams().setSpeed(speed)
                }
                result.success(null)
            }

            // ── Waveform extraction ───────────────────────────────────────
            "extractWaveform" -> {
                val path        = call.argument<String>("path") ?: return result.error("ARGS", "path required", null)
                val sampleCount = call.argument<Int>("sampleCount") ?: 200

                Thread {
                    try {
                        val samples = extractWaveformFromFile(path, sampleCount)
                        channel.invokeMethod("_waveformResult", null) // not used
                        result.success(samples)
                    } catch (e: Exception) {
                        result.error("WAVEFORM_ERROR", e.message, null)
                    }
                }.start()
            }

            else -> result.notImplemented()
        }
    }

    // ── MediaRecorder setup ───────────────────────────────────────────────

    private fun startMediaRecorder(
        path: String, format: String,
        sampleRate: Int, bitRate: Int, channels: Int
    ) {
        mediaRecorder?.release()
        mediaRecorder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            MediaRecorder(context!!)
        } else {
            @Suppress("DEPRECATION")
            MediaRecorder()
        }

        mediaRecorder?.apply {
            setAudioSource(MediaRecorder.AudioSource.MIC)
            setOutputFormat(when (format) {
                "wav"  -> MediaRecorder.OutputFormat.DEFAULT
                "ogg"  -> MediaRecorder.OutputFormat.OGG
                else   -> MediaRecorder.OutputFormat.MPEG_4
            })
            setAudioEncoder(when (format) {
                "ogg"  -> MediaRecorder.AudioEncoder.VORBIS
                else   -> MediaRecorder.AudioEncoder.AAC
            })
            setAudioSamplingRate(sampleRate)
            setAudioEncodingBitRate(bitRate)
            setAudioChannels(channels)
            setOutputFile(path)
            prepare()
            start()
        }
    }

    // ── File path builder ─────────────────────────────────────────────────

    private fun buildOutputPath(format: String, outputDir: String?, fileName: String?): String {
        val dir = outputDir?.let { File(it) }
            ?: context!!.cacheDir
        dir.mkdirs()
        val name = fileName ?: "recording_${SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(Date())}"
        return "${dir.absolutePath}/$name.$format"
    }

    // ── Waveform extraction from audio file ───────────────────────────────

    private fun extractWaveformFromFile(path: String, sampleCount: Int): List<Double> {
        val extractor = MediaExtractor()
        extractor.setDataSource(path)

        // Find audio track
        var audioTrackIndex = -1
        var format: MediaFormat? = null
        for (i in 0 until extractor.trackCount) {
            val f = extractor.getTrackFormat(i)
            if (f.getString(MediaFormat.KEY_MIME)?.startsWith("audio/") == true) {
                audioTrackIndex = i
                format = f
                break
            }
        }

        if (audioTrackIndex < 0 || format == null) {
            extractor.release()
            return List(sampleCount) { 0.0 }
        }

        extractor.selectTrack(audioTrackIndex)

        val codec = MediaCodec.createDecoderByType(
            format.getString(MediaFormat.KEY_MIME)!!
        )
        codec.configure(format, null, null, 0)
        codec.start()

        val rawSamples = mutableListOf<Short>()
        val bufferInfo = MediaCodec.BufferInfo()
        var inputDone = false
        var outputDone = false

        while (!outputDone) {
            if (!inputDone) {
                val inputIdx = codec.dequeueInputBuffer(10000)
                if (inputIdx >= 0) {
                    val buf = codec.getInputBuffer(inputIdx)!!
                    val size = extractor.readSampleData(buf, 0)
                    if (size < 0) {
                        codec.queueInputBuffer(inputIdx, 0, 0, 0, MediaCodec.BUFFER_FLAG_END_OF_STREAM)
                        inputDone = true
                    } else {
                        codec.queueInputBuffer(inputIdx, 0, size, extractor.sampleTime, 0)
                        extractor.advance()
                    }
                }
            }

            val outputIdx = codec.dequeueOutputBuffer(bufferInfo, 10000)
            if (outputIdx >= 0) {
                val buf = codec.getOutputBuffer(outputIdx)!!
                val shorts = buf.order(ByteOrder.LITTLE_ENDIAN).asShortBuffer()
                while (shorts.hasRemaining()) rawSamples.add(shorts.get())
                codec.releaseOutputBuffer(outputIdx, false)
                if (bufferInfo.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0) {
                    outputDone = true
                }
            }
        }

        codec.stop()
        codec.release()
        extractor.release()

        // Bucket into sampleCount chunks, compute RMS of each
        if (rawSamples.isEmpty()) return List(sampleCount) { 0.0 }
        val bucketSize = maxOf(1, rawSamples.size / sampleCount)
        return (0 until sampleCount).map { i ->
            val start = i * bucketSize
            val end   = minOf(start + bucketSize, rawSamples.size)
            if (start >= rawSamples.size) return@map 0.0
            var sum = 0.0
            for (j in start until end) sum += rawSamples[j] * rawSamples[j].toDouble()
            val rms = sqrt(sum / (end - start))
            (rms / Short.MAX_VALUE).coerceIn(0.0, 1.0)
        }
    }
}

File: ios/Classes/AudioWaveformPlugin.swift
import Flutter
import UIKit
import AVFoundation
import Accelerate

public class AudioWaveformPlugin: NSObject, FlutterPlugin {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "audio_waveform_recorder",
            binaryMessenger: registrar.messenger()
        )
        let instance = AudioWaveformPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    // ── State ─────────────────────────────────────────────────────────────
    private var audioRecorder: AVAudioRecorder?
    private var audioPlayer:   AVAudioPlayer?
    private var currentFilePath: String?
    private var recordingStartTime: Date?

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]

        switch call.method {

        // ── Permissions ───────────────────────────────────────────────────
        case "hasMicPermission":
            result(AVAudioSession.sharedInstance().recordPermission == .granted)

        case "requestMicPermission":
            AVAudioSession.sharedInstance().requestRecordPermission { granted in
                DispatchQueue.main.async { result(granted) }
            }

        // ── Recording ─────────────────────────────────────────────────────
        case "startRecording":
            let format      = args["format"]     as? String ?? "m4a"
            let sampleRate  = args["sampleRate"]  as? Double ?? 44100.0
            let bitRate     = args["bitRate"]     as? Int    ?? 128000
            let channels    = args["channels"]    as? Int    ?? 1
            let outputDir   = args["outputDir"]   as? String
            let fileName    = args["fileName"]    as? String

            do {
                let path = buildOutputPath(format: format, outputDir: outputDir, fileName: fileName)
                try startAudioRecorder(
                    path: path, format: format,
                    sampleRate: sampleRate, bitRate: bitRate, channels: channels
                )
                currentFilePath = path
                recordingStartTime = Date()
                result(path)
            } catch {
                result(FlutterError(code: "RECORD_ERROR", message: error.localizedDescription, details: nil))
            }

        case "pauseRecording":
            audioRecorder?.pause()
            result(nil)

        case "resumeRecording":
            audioRecorder?.record()
            result(nil)

        case "stopRecording":
            audioRecorder?.stop()
            let path     = currentFilePath ?? ""
            let url      = URL(fileURLWithPath: path)
            let size     = (try? FileManager.default.attributesOfItem(atPath: path)[.size] as? Int) ?? 0
            let duration = Int((Date().timeIntervalSince(recordingStartTime ?? Date())) * 1000)

            try? AVAudioSession.sharedInstance().setActive(false)
            audioRecorder = nil
            currentFilePath = nil

            result(["path": path, "durationMs": duration, "sizeBytes": size])

        case "cancelRecording":
            audioRecorder?.stop()
            audioRecorder = nil
            if let path = currentFilePath {
                try? FileManager.default.removeItem(atPath: path)
            }
            currentFilePath = nil
            try? AVAudioSession.sharedInstance().setActive(false)
            result(nil)

        case "getAmplitude":
            guard let recorder = audioRecorder else { result(0.0); return }
            recorder.updateMeters()
            // averagePower is in dBFS (-160 to 0). Convert to 0.0–1.0
            let dB      = recorder.averagePower(forChannel: 0)
            let minDB   = -60.0 as Float
            let normalised: Double
            if dB < minDB {
                normalised = 0.0
            } else if dB >= 0 {
                normalised = 1.0
            } else {
                normalised = Double((dB - minDB) / (0 - minDB))
            }
            result(normalised)

        // ── Playback ──────────────────────────────────────────────────────
        case "loadAudio":
            guard let path = args["path"] as? String else {
                result(FlutterError(code: "ARGS", message: "path required", details: nil))
                return
            }
            do {
                try AVAudioSession.sharedInstance().setCategory(.playback)
                try AVAudioSession.sharedInstance().setActive(true)
                audioPlayer = try AVAudioPlayer(contentsOf: URL(fileURLWithPath: path))
                audioPlayer?.prepareToPlay()
                let durationMs = Int((audioPlayer?.duration ?? 0) * 1000)
                result(["durationMs": durationMs])
            } catch {
                result(FlutterError(code: "LOAD_ERROR", message: error.localizedDescription, details: nil))
            }

        case "playAudio":
            audioPlayer?.play()
            result(nil)

        case "pauseAudio":
            audioPlayer?.pause()
            result(nil)

        case "stopAudio":
            audioPlayer?.stop()
            audioPlayer?.currentTime = 0
            result(nil)

        case "seekTo":
            let ms = args["positionMs"] as? Int ?? 0
            audioPlayer?.currentTime = Double(ms) / 1000.0
            result(nil)

        case "getPlaybackPosition":
            let ms = Int((audioPlayer?.currentTime ?? 0) * 1000)
            result(ms)

        case "setVolume":
            let vol = args["volume"] as? Float ?? 1.0
            audioPlayer?.volume = vol
            result(nil)

        case "setSpeed":
            let speed = args["speed"] as? Float ?? 1.0
            audioPlayer?.rate = speed
            result(nil)

        // ── Waveform extraction ───────────────────────────────────────────
        case "extractWaveform":
            guard let path = args["path"] as? String else {
                result(FlutterError(code: "ARGS", message: "path required", details: nil))
                return
            }
            let sampleCount = args["sampleCount"] as? Int ?? 200

            DispatchQueue.global(qos: .userInitiated).async {
                let samples = self.extractWaveform(from: path, sampleCount: sampleCount)
                DispatchQueue.main.async { result(samples) }
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // ── AVAudioRecorder setup ─────────────────────────────────────────────

    private func startAudioRecorder(
        path: String, format: String,
        sampleRate: Double, bitRate: Int, channels: Int
    ) throws {
        try AVAudioSession.sharedInstance().setCategory(.playAndRecord, options: [.defaultToSpeaker])
        try AVAudioSession.sharedInstance().setActive(true)

        let formatID: AudioFormatID
        let ext = format.lowercased()
        if ext == "wav" {
            formatID = kAudioFormatLinearPCM
        } else {
            formatID = kAudioFormatMPEG4AAC
        }

        let settings: [String: Any] = [
            AVFormatIDKey:              formatID,
            AVSampleRateKey:            sampleRate,
            AVNumberOfChannelsKey:      channels,
            AVEncoderBitRateKey:        bitRate,
            AVEncoderAudioQualityKey:   AVAudioQuality.high.rawValue,
            AVLinearPCMBitDepthKey:     16,
            AVLinearPCMIsBigEndianKey:  false,
            AVLinearPCMIsFloatKey:      false,
        ]

        audioRecorder = try AVAudioRecorder(
            url: URL(fileURLWithPath: path),
            settings: settings
        )
        audioRecorder?.isMeteringEnabled = true
        audioRecorder?.record()
    }

    // ── Output path builder ───────────────────────────────────────────────

    private func buildOutputPath(format: String, outputDir: String?, fileName: String?) -> String {
        let dir: URL
        if let d = outputDir {
            dir = URL(fileURLWithPath: d)
        } else {
            dir = FileManager.default.temporaryDirectory
        }
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

        let name = fileName ?? "recording_\(Int(Date().timeIntervalSince1970))"
        return dir.appendingPathComponent("\(name).\(format)").path
    }

    // ── Waveform extraction ───────────────────────────────────────────────

    private func extractWaveform(from path: String, sampleCount: Int) -> [Double] {
        guard let file = try? AVAudioFile(forReading: URL(fileURLWithPath: path)),
              let format = AVAudioFormat(
                  commonFormat: .pcmFormatFloat32,
                  sampleRate: file.fileFormat.sampleRate,
                  channels: 1,
                  interleaved: false
              ) else {
            return Array(repeating: 0.0, count: sampleCount)
        }

        let frameCount = AVAudioFrameCount(file.length)
        guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount) else {
            return Array(repeating: 0.0, count: sampleCount)
        }

        do {
            try file.read(into: buffer)
        } catch {
            return Array(repeating: 0.0, count: sampleCount)
        }

        guard let floatData = buffer.floatChannelData?[0] else {
            return Array(repeating: 0.0, count: sampleCount)
        }

        let totalSamples = Int(buffer.frameLength)
        let bucketSize   = max(1, totalSamples / sampleCount)

        var result = [Double]()
        for i in 0..<sampleCount {
            let start = i * bucketSize
            let end   = min(start + bucketSize, totalSamples)
            guard start < totalSamples else { result.append(0.0); continue }

            // RMS of the bucket using Accelerate
            var rms: Float = 0
            vDSP_rmsqv(floatData + start, 1, &rms, vDSP_Length(end - start))
            result.append(Double(min(rms, 1.0)))
        }

        return result
    }
}
